feat/palette-json-generator-webapp

klown opened the pull request:
A web-app for entering glosses, BCI AV IDs, and svg builder strings to produce and save a palette.

The main point is to make it easier than the previous way to make palettes by using  text editor to edit the JSON description file directly.  The previous way was time consuming and very error prone.  The web-app approach allows most of the generation of the JSON to be automated, and allows the user to later tweak the JSON file as necessary.

This branch was originally created based on the code for PR #21, but incorporated @cindyli's script into user interface.  It's not clear if this should be directly merged with the main branch or merged into the feat/palette-json-generator script branch first.

cindyli commented on apps/palette-generator/palette-generator.html:
```diff
@@ -0,0 +1,72 @@
+      List the items in each row, one line per row.  Each row consists of a set
+      of glosses separated by a space.
+    </p>
+      <label for="glossInput">Glosses:</label><br>
```
The label for the text area to receive BCI-AV-ID, glosses, SVGs etc should no longer be "Glosses" since it accepts a lot more than glosses.

klown replied:
Right, I'll come up with a better name.

cindyli commented on apps/palette-generator/paletteJsonGenerator.ts:
```diff
@@ -0,0 +1,252 @@
+const BLANK_CELL_LABEL = "BLANK";
+const SVG_PREFIX = "SVG:";
+const SVG_SUFFIX = ":SVG";
+const LABEL_MARKER = "LABEL:";
```
When I input this line in the "Glosses" text area:

```
BLANK 18267LABEL:aaa induction
```
to alter the label for a symbol. I got this error in the error section: "BciAvId not found for label: induction". Moving the "LABEL:" marker to the last "induction" gloss reports the same error.

klown replied:
That string specifies three cells:  a blank one, one for a plus sign labelled "aaa" and a cell whose gloss is "induction" or contains the word induction.  "Induction" does not match any gloss and that is why there is an error.  Checking blissary.com with "induction" gives "No results found."  The result is as expected.  Moving the LABEL marker won't change the result.

There was some text in the README explaining that spaces are used as a separator to indicate the string to use for a cell.  I've added more warnings to that effect including that LABEL: strings cannot contain spaces and to use an underscore if a space is desired.  Specifically, if the label for the plus sign was supposed to be "aaa induction", then you write it this way:

`BLANK 18267LABEL:aaa_induction`

That gives two cells, a blank and a plus with the label "aaa induction".

cindyli replied:
You are right. The error is reported correctly. I mistakenly thought it was related to the "LABEL" marker.

klown commented on README.md:
```diff
@@ -132,3 +136,8 @@ command" and "Build output directory" settings:
+## Utility Scripts
+
+- [Load a document into a vector databbase (`scripts/loadDocIntoVectorDb.js`)](./scripts/loadDocIntoVectorDb.js)
+- [Generate a palette JSON file (`scripts/paletteJsonGenerator.js`)](./scripts/loadDocIntoVectorDb.js)
```
Yup.  I'll remove it.

klown commented on apps/palette-generator/README.md:
```diff
@@ -0,0 +1,141 @@
+
+Once the development server is running, open this `localhost` url from within a
+browser:
+[`http://localhost:5173/apps/palette-generator/palette-generator.html`](http://localhost:5173/apps/palette-generator/palette-generator.html)
```
Good idea.

cindyli replied:
Since the entry page has been renamed to "index.html", the link and text should be changed to remove "palette-generator.html".

klown commented on apps/palette-generator/README.md:
```diff
@@ -0,0 +1,141 @@
+row of the palette cells.  Spaces are used for separate cells so the text used
+to specify a cell cannot contain any spaces.
+
+There are three types of input allowed here:
```
Yes.

klown commented on apps/palette-generator/README.md:
```diff
@@ -0,0 +1,141 @@
+the label "cross hatches"; and "23641LABEL:ruin" will give "ruin" as the label
+instead of the official gloss "ruin,wreck,wreckage\_(building)\_(1)".
+
+The SVG string is marked by "SVG:" at the beginning of the string and ":SVG" at
```
I've added two examples. But, I am not going to explain the full syntax of svg builder strings.  That should be in the svg-builder project, but there is no documentation there.  And, I don't know the full specification.  :-(

klown commented on apps/palette-generator/README.md:
```diff
@@ -0,0 +1,141 @@
+### Save Palette
+
+Clicking the "Save palette" button will export a palette definition JSON file
+and save it into the user's "Downloads" folder.  The name of the file will be
```
Good question. The code itself does not include the name "Downloads" or the name of any other folder.  An url is created that the browser understands, but is meaningless to me. The file ends up in the folder that has been set as the browser's download folder.  Users can set this to something other than "Downloads", in which case the palette will be saved based on that browser setting.

I will reword it to say, '.. and saved to disk where the browser saves file downloads, typically the "Downloads" folder.'

klown commented on apps/palette-generator/palette-generator.ts:
```diff
@@ -0,0 +1,187 @@
+// Initialize any globals used elsewhere in the code.
+await initAdaptivePaletteGlobals("paletteDisplay");
+await fetchBlissGlossJson();
+let currentPaletteName = "";
```
I found a bug here and fixed it, but I have not addressed your comment fully:

> The save button doesn't save the given palette name.

It does after the fix, but there is still an edge case where (1) the palette is generated, (2) the user changes the name in the text field, and then (3) clicks the "Save" button.  In that case, the file is saved according the name in the text field when the palette was generated, not the just-changed text field  For this case, there needs to be an event handler for watching changes in the text field and adjusting various objects.  I'm working on that.

> The clear button doesn't clear the palette name too.

I don't think it should clear the name in the text field.  If a palette has been generated, but there are mistakes, and all the user wants to do is clear it, fix the mistake(s), and re-generate, then they should not have to re-enter the name. Their intent is to change the contents of the palette, not the name.  When they do want to change the name, then it's relatively simple to click on the text field and give it a new name.

cindyli replied:
> I don't think it should clear the name in the text field.

Thanks for the clarification, and I agree.

klown commented on apps/palette-generator/paletteJsonGenerator.ts:
```diff
@@ -0,0 +1,252 @@
+const LABEL_MARKER = "LABEL:";
+
+// Configurable options -- ideally provided by the UI
+const palette_name = "No name Palette";
```
Good idea.  I will get the contents of the name text field and pass it via the generator function.

klown commented on apps/palette-generator/paletteJsonGenerator.ts:
```diff
@@ -0,0 +1,252 @@
+const type = "ActionBmwCodeCell";
+
+// Special encodings for labels that don't follow the standard mapping
+const specialEncodings = {
```
I'm inclined to remove it now.  But, just in case, what did you intend it for? Do the SVG: and LABEL: features cover that purpose?

cindyli replied:
I agree this special handling can be removed.

cindyli replied:
Remove the definition and the use of `specialEncodings`?

cindyli commented on apps/palette-generator/README.md:
```diff
@@ -0,0 +1,141 @@
+strings that are parsed by the svg builder.  If the string "LABEL:*label\_text"
+occurs immediately before the closing ":SVG" marker, it will be used as the
+label for the cell, and is truncated from the rest of the string before the
+remainder is sent to the builder.  The rules for the syntax of this "LABEL" is
```
I tried with using "LABEL" with "SVG" marker. This string `SVG:14183,"/",25777,"/","W8W:0,8":SVGLABEL:aaa` leads to an error: "BciAvId not found for label: SVG:14183,"/",25777,"/","W8W:0,8":SVG". Did I combine the "SVG" and "LABEL" correctly?

klown replied:
Quoting from the README with emphasis:
> If the string "LABEL:*label\_text" occurs _immediately before the closing ":SVG" marker_, 

So your example should be `SVG:14183,"/",25777,"/","W8W:0,8","LABEL:aaa":SVG`

However, that's different from the two other cases of an ID or a gloss-string, where the `LABEL:` is appended to the ID or string.  I think it should be the same regardless of the input.  I have re-written the processing so that your example works, and the old way is now illegal.

cindyli commented:
It would be helpful for users if you could add a small helper section next to the "Glosses" text field on the app interface. This area could provide quick explanations or examples of available markers like SVG, LABEL, and BLANK. I found I often doubt if I'm using these markers correctly and have to check README frequently.

klown commented on apps/palette-generator/paletteJsonGenerator.ts:
```diff
@@ -0,0 +1,252 @@
+export async function fetchBlissGlossJson () {
+  // Read and parse the Bliss gloss JSON file
+  try {
+    const fetchResponse = await fetch("https://raw.githubusercontent.com/cindyli/baby-bliss-bot/refs/heads/feat/bmw/data/bliss_symbol_explanations.json");
```
Good idea. I found a way to merge that one json file across the two repositories so the history is maintained.

klown commented on apps/palette-generator/README.md:
````diff
@@ -0,0 +1,141 @@
+
+```text
+cloth
+    13365
````
I tried that out and it looks better -- thanks!

klown commented:
@cindyli  wrote:

> It would be helpful for users if you could add a small helper section next to the "Glosses" text field on the app interface. This area could provide quick explanations or examples of available markers like SVG, LABEL, and BLANK. I found I often doubt if I'm using these markers correctly and have to check README frequently.

I added one but in a way that allows users to pop it open as needed and then hide it.  I used the `details/summary` technique.  I have written and re-written the text for it, and am still not satisfied.  But, I have to stop at some point, so here it is.

I think I addressed all of your comments -- ready for another review, thanks.

cindyli commented on apps/palette-generator/palette-generator.ts:
```diff
@@ -0,0 +1,223 @@
+      const dd = document.createElement("dd");
+      dl.append(dd);
+      dd.innerText = `${match.bciAvId}: ${match.label}`;
+      //       dd = document.createElement("dd");
+      //       dl.append(dd);
+      //       dd.innerText = match.label;
```
Remove the commented code?

klown replied:
Yes, no longer needed.

cindyli commented on apps/palette-generator/paletteJsonGenerator.ts:
```diff
@@ -0,0 +1,241 @@
+  if (matches.length === 0) {
+    throw new Error(`BciAvId not found for label: ${label}`);
+  }
+  console.log("");
```
Shall this line be removed?

klown replied:
It was meant to output a blank line to separate the matches logged on the console.  But in FF it insists on printing `<empty string>` which is not a blank line.  I switched it to `console.log("\n");` and that does print a blank line.  Then I took it out completely, and the logged matches are readable even without a separating blank line.  So, no it's been removed.

klown commented:
@cindyli, I've addressed you last comments.  I also added a new `<select>` element to allow users to choose the cell type.   However one type, the `ContentBmwEncoding` cell type, does not render well.  I am investigating.

Nonetheless, ready for another review.

cindyli commented:
The pull request looks great. I will wait for the fix of the rendering of the `ContentBmwEncoding` cell type.

klown commented:
I did manage to render `ContentBmwEncoding` by setting  up the `changeEncodingContents` signal correctly.  However, since the display area is a palette and not an content area, each cell is treated as a content area and all of the bliss symbols are shown in every cell.  It looks bizarre.

Then I realized that the palette json file to be saved should ever have cells of type `ContentBmwEncoding`.  So, since this is about _palette_ generation and not _content_ generation, the select is set up so that ContentBmwEncoding` is unavailable.

Let me know what you think.

cindyli commented:
> Then I realized that the palette json file to be saved should never have cells of type `ContentBmwEncoding`. So, since this is about _palette_ generation and not _content_ generation, the select is set up so that `ContentBmwEncoding` is unavailable.

I agree this is fine.
